chapter_attention-mechanisms/transformer.md

```diff
@@ -122,6 +122,19 @@ import torch
 from torch import nn
 ```
 
+```{.python .input}
+#@tab tensorflow
+from d2l import tensorflow as d2l
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+gpus = tf.config.list_physical_devices("GPU")
+if gpus:
+    for gpu in gpus:
+        tf.config.experimental.set_memory_growth(gpu, True)
```

Comment:
why do we need this?

Reply:
In TF-2.3, if we don't add these lines, we get an error in the following code block in the same section 10.7.3.
```
ln = tf.keras.layers.LayerNormalization()
bn = tf.keras.layers.BatchNormalization()
X = tf.constant([[1, 2], [2, 3]], dtype=tf.float32)
print('layer norm:', ln(X), '\nbatch norm:', bn(X, training=True))
```
Surprisingly, if we run the same code block in TF-2.4 or TF-2.5, it runs without any issue. Initially, I had not added these lines, but because I got an error while build process, I had to add it in this commit [54caa6c](https://github.com/d2l-ai/d2l-en/pull/1768/commits/54caa6cd7f1f832b16563b5bc94310a8c41a8a49). Why this happens? I have not the slightest of idea.

Reply:
What errors are you getting if this is not specified?

Reply:
To reproduce the error, I created [1788](https://github.com/d2l-ai/d2l-en/pull/1788). Upon close inspection, I found that the error is somewhere else other than what I had mentioned above. Please find the full log of the error. @terrytangyuan @astonzhang 
```
+ d2lbook build eval --tab tensorflow

[d2lbook:config.py:L12] INFO   Load configure from config.ini

[d2lbook:build.py:L143] INFO   3 notebooks are outdated

[d2lbook:build.py:L145] INFO   [1] ./chapter_convolutional-neural-networks/lenet.md

[d2lbook:build.py:L145] INFO   [2] ./chapter_attention-mechanisms/transformer.md

[d2lbook:build.py:L145] INFO   [3] ./chapter_linear-networks/softmax-regression-scratch.md

[d2lbook:build.py:L204] INFO   Copying resources "img/ d2l/ d2l.bib setup.py" from . to _build/eval_tensorflow

[d2lbook:build.py:L149] INFO   Evaluating notebooks in parallel with 4 CPU workers and 4 GPU workers

[d2lbook:resource.py:L195] INFO   Starting task "Evaluating ./chapter_convolutional-neural-networks/lenet.md" on CPU [3], GPU [1]

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 0 done, 2 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_convolutional-neural-networks/lenet.md" on CPU [3], GPU [1] is running for 00:00:00

[d2lbook:execute.py:L404] INFO   Executing notebook with kernel: python

[d2lbook:resource.py:L195] INFO   Starting task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3]

[d2lbook:resource.py:L158] INFO     Status: 2 running tasks, 0 done, 1 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_convolutional-neural-networks/lenet.md" on CPU [3], GPU [1] is running for 00:00:02

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:00:00

[d2lbook:execute.py:L404] INFO   Executing notebook with kernel: python

[d2lbook:resource.py:L195] INFO   Starting task "Evaluating ./chapter_linear-networks/softmax-regression-scratch.md" on CPU [0]

[d2lbook:resource.py:L158] INFO     Status: 3 running tasks, 0 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_convolutional-neural-networks/lenet.md" on CPU [3], GPU [1] is running for 00:00:04

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:00:02

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_linear-networks/softmax-regression-scratch.md" on CPU [0] is running for 00:00:00

[d2lbook:execute.py:L404] INFO   Executing notebook with kernel: python

[d2lbook:resource.py:L218] ERROR  Task "Evaluating ./chapter_convolutional-neural-networks/lenet.md" on CPU [3], GPU [1] exited with error: An error occurred while executing the following cell:

------------------

X = tf.random.uniform((1, 28, 28, 1))

for layer in net().layers:

    X = layer(X)

    print(layer.__class__.__name__, 'output shape: \t', X.shape)

------------------



---------------------------------------------------------------------------

InternalError                             Traceback (most recent call last)

<ipython-input-2-f64dd12b72ac> in <module>

      1 X = tf.random.uniform((1, 28, 28, 1))

      2 for layer in net().layers:

----> 3     X = layer(X)

      4     print(layer.__class__.__name__, 'output shape: \t', X.shape)



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)

    983 

    984         with ops.enable_auto_cast_variables(self._compute_dtype_object):

--> 985           outputs = call_fn(inputs, *args, **kwargs)

    986 

    987         if self._activity_regularizer:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py in call(self, inputs)

   1191 

   1192   def call(self, inputs):

-> 1193     return core_ops.dense(

   1194         inputs,

   1195         self.kernel,



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/ops/core.py in dense(inputs, kernel, bias, activation, dtype)

     51       outputs = sparse_ops.sparse_tensor_dense_matmul(inputs, kernel)

     52     else:

---> 53       outputs = gen_math_ops.mat_mul(inputs, kernel)

     54   # Broadcast kernel to inputs.

     55   else:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py in mat_mul(a, b, transpose_a, transpose_b, name)

   5622       return _result

   5623     except _core._NotOkStatusException as e:

-> 5624       _ops.raise_from_not_ok_status(e, name)

   5625     except _core._FallbackException:

   5626       pass



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)

   6841   message = e.message + (" name: " + name if name is not None else "")

   6842   # pylint: disable=protected-access

-> 6843   six.raise_from(core._status_to_exception(e.code, message), None)

   6844   # pylint: enable=protected-access

   6845 



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/six.py in raise_from(value, from_value)



InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]

InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]



Traceback (most recent call last):

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/resource.py", line 68, in run

    mp.Process.run(self)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/multiprocessing/process.py", line 108, in run

    self._target(*self._args, **self._kwargs)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/resource.py", line 255, in _target

    return target(*args)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/build.py", line 507, in _job

    notedown.run(nb, timeout)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/notedown/notedown.py", line 48, in run

    notebook, resources = executor.preprocess(notebook, resources={})

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/execute.py", line 405, in preprocess

    nb, resources = super(ExecutePreprocessor, self).preprocess(nb, resources)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/base.py", line 69, in preprocess

    nb.cells[index], resources = self.preprocess_cell(cell, resources, index)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/execute.py", line 448, in preprocess_cell

    raise CellExecutionError.from_cell_and_msg(cell, out)

nbconvert.preprocessors.execute.CellExecutionError: An error occurred while executing the following cell:

------------------

X = tf.random.uniform((1, 28, 28, 1))

for layer in net().layers:

    X = layer(X)

    print(layer.__class__.__name__, 'output shape: \t', X.shape)

------------------



---------------------------------------------------------------------------

InternalError                             Traceback (most recent call last)

<ipython-input-2-f64dd12b72ac> in <module>

      1 X = tf.random.uniform((1, 28, 28, 1))

      2 for layer in net().layers:

----> 3     X = layer(X)

      4     print(layer.__class__.__name__, 'output shape: \t', X.shape)



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)

    983 

    984         with ops.enable_auto_cast_variables(self._compute_dtype_object):

--> 985           outputs = call_fn(inputs, *args, **kwargs)

    986 

    987         if self._activity_regularizer:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py in call(self, inputs)

   1191 

   1192   def call(self, inputs):

-> 1193     return core_ops.dense(

   1194         inputs,

   1195         self.kernel,



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/ops/core.py in dense(inputs, kernel, bias, activation, dtype)

     51       outputs = sparse_ops.sparse_tensor_dense_matmul(inputs, kernel)

     52     else:

---> 53       outputs = gen_math_ops.mat_mul(inputs, kernel)

     54   # Broadcast kernel to inputs.

     55   else:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py in mat_mul(a, b, transpose_a, transpose_b, name)

   5622       return _result

   5623     except _core._NotOkStatusException as e:

-> 5624       _ops.raise_from_not_ok_status(e, name)

   5625     except _core._FallbackException:

   5626       pass



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)

   6841   message = e.message + (" name: " + name if name is not None else "")

   6842   # pylint: disable=protected-access

-> 6843   six.raise_from(core._status_to_exception(e.code, message), None)

   6844   # pylint: enable=protected-access

   6845 



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/six.py in raise_from(value, from_value)



InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]

InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]





[d2lbook:resource.py:L222] INFO   Task "Evaluating ./chapter_linear-networks/softmax-regression-scratch.md" on CPU [0] is finished in 00:00:23

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:01:04

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:02:05

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:03:06

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:04:07

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:05:08

[d2lbook:resource.py:L158] INFO     Status: 1 running tasks, 2 done, 0 not started

[d2lbook:resource.py:L163] INFO       - Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is running for 00:06:09

[d2lbook:resource.py:L222] INFO   Task "Evaluating ./chapter_attention-mechanisms/transformer.md" on CPU [1], GPU [3] is finished in 00:06:15

[d2lbook:resource.py:L141] INFO   All 3 tasks are done, sorting by runtime:

[d2lbook:resource.py:L147] INFO     - 00:00:08 on CPU [3], GPU [1] for Evaluating ./chapter_convolutional-neural-networks/lenet.md

[d2lbook:resource.py:L147] INFO     - 00:00:23 on CPU [0] for Evaluating ./chapter_linear-networks/softmax-regression-scratch.md

[d2lbook:resource.py:L147] INFO     - 00:06:15 on CPU [1], GPU [3] for Evaluating ./chapter_attention-mechanisms/transformer.md

Traceback (most recent call last):

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/bin/d2lbook", line 8, in <module>

    sys.exit(main())

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/main.py", line 25, in main

    commands[args.command[0]]()

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/build.py", line 40, in build

    getattr(builder, cmd)()

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/build.py", line 52, in warp

    func(self)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/build.py", line 159, in eval

    assert not scheduler.failed_tasks, scheduler.error_message

AssertionError: 1 notebooks are failed to evaluate:



Task Evaluating ./chapter_convolutional-neural-networks/lenet.md exited with error: An error occurred while executing the following cell:

------------------

X = tf.random.uniform((1, 28, 28, 1))

for layer in net().layers:

    X = layer(X)

    print(layer.__class__.__name__, 'output shape: \t', X.shape)

------------------



---------------------------------------------------------------------------

InternalError                             Traceback (most recent call last)

<ipython-input-2-f64dd12b72ac> in <module>

      1 X = tf.random.uniform((1, 28, 28, 1))

      2 for layer in net().layers:

----> 3     X = layer(X)

      4     print(layer.__class__.__name__, 'output shape: \t', X.shape)



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)

    983 

    984         with ops.enable_auto_cast_variables(self._compute_dtype_object):

--> 985           outputs = call_fn(inputs, *args, **kwargs)

    986 

    987         if self._activity_regularizer:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py in call(self, inputs)

   1191 

   1192   def call(self, inputs):

-> 1193     return core_ops.dense(

   1194         inputs,

   1195         self.kernel,



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/ops/core.py in dense(inputs, kernel, bias, activation, dtype)

     51       outputs = sparse_ops.sparse_tensor_dense_matmul(inputs, kernel)

     52     else:

---> 53       outputs = gen_math_ops.mat_mul(inputs, kernel)

     54   # Broadcast kernel to inputs.

     55   else:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py in mat_mul(a, b, transpose_a, transpose_b, name)

   5622       return _result

   5623     except _core._NotOkStatusException as e:

-> 5624       _ops.raise_from_not_ok_status(e, name)

   5625     except _core._FallbackException:

   5626       pass



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)

   6841   message = e.message + (" name: " + name if name is not None else "")

   6842   # pylint: disable=protected-access

-> 6843   six.raise_from(core._status_to_exception(e.code, message), None)

   6844   # pylint: enable=protected-access

   6845 



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/six.py in raise_from(value, from_value)



InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]

InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]





Traceback (most recent call last):

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/resource.py", line 68, in run

    mp.Process.run(self)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/multiprocessing/process.py", line 108, in run

    self._target(*self._args, **self._kwargs)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/resource.py", line 255, in _target

    return target(*args)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/d2lbook/build.py", line 507, in _job

    notedown.run(nb, timeout)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/notedown/notedown.py", line 48, in run

    notebook, resources = executor.preprocess(notebook, resources={})

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/execute.py", line 405, in preprocess

    nb, resources = super(ExecutePreprocessor, self).preprocess(nb, resources)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/base.py", line 69, in preprocess

    nb.cells[index], resources = self.preprocess_cell(cell, resources, index)

  File "/var/lib/jenkins/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/nbconvert/preprocessors/execute.py", line 448, in preprocess_cell

    raise CellExecutionError.from_cell_and_msg(cell, out)

nbconvert.preprocessors.execute.CellExecutionError: An error occurred while executing the following cell:

------------------

X = tf.random.uniform((1, 28, 28, 1))

for layer in net().layers:

    X = layer(X)

    print(layer.__class__.__name__, 'output shape: \t', X.shape)

------------------



---------------------------------------------------------------------------

InternalError                             Traceback (most recent call last)

<ipython-input-2-f64dd12b72ac> in <module>

      1 X = tf.random.uniform((1, 28, 28, 1))

      2 for layer in net().layers:

----> 3     X = layer(X)

      4     print(layer.__class__.__name__, 'output shape: \t', X.shape)



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)

    983 

    984         with ops.enable_auto_cast_variables(self._compute_dtype_object):

--> 985           outputs = call_fn(inputs, *args, **kwargs)

    986 

    987         if self._activity_regularizer:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py in call(self, inputs)

   1191 

   1192   def call(self, inputs):

-> 1193     return core_ops.dense(

   1194         inputs,

   1195         self.kernel,



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/keras/layers/ops/core.py in dense(inputs, kernel, bias, activation, dtype)

     51       outputs = sparse_ops.sparse_tensor_dense_matmul(inputs, kernel)

     52     else:

---> 53       outputs = gen_math_ops.mat_mul(inputs, kernel)

     54   # Broadcast kernel to inputs.

     55   else:



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py in mat_mul(a, b, transpose_a, transpose_b, name)

   5622       return _result

   5623     except _core._NotOkStatusException as e:

-> 5624       _ops.raise_from_not_ok_status(e, name)

   5625     except _core._FallbackException:

   5626       pass



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)

   6841   message = e.message + (" name: " + name if name is not None else "")

   6842   # pylint: disable=protected-access

-> 6843   six.raise_from(core._status_to_exception(e.code, message), None)

   6844   # pylint: enable=protected-access

   6845 



~/miniconda3/envs/d2l-en-master-0/lib/python3.8/site-packages/six.py in raise_from(value, from_value)



InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]

InternalError: Blas GEMM launch failed : a.shape=(1, 400), b.shape=(400, 120), m=1, n=120, k=400 [Op:MatMul]





script returned exit code 1
```